packages/eslint-config/nextjs.js

```diff
@@ -19,5 +19,42 @@ module.exports = {
                 ],
             },
         ],
+        "no-restricted-syntax": [
+            "error",
+            {
+                selector:
+                    "MemberExpression[type=MemberExpression][object.type=MemberExpression][object.object.type=Identifier][object.object.name=process][object.property.type=Identifier][object.property.name=env][property.type=Identifier][property.name=/^NEXT_PUBLIC/]",
+                message: "Avoid using NEXT_PUBLIC_ environment variables. Use SiteConfig or a custom Context instead",
+            },
+        ],
     },
+    overrides: [
+        {
+            files: ["next.config.js"],
+            rules: {
+                "no-restricted-syntax": [
+                    "error",
+                    {
+                        selector:
+                            "MemberExpression[type=MemberExpression][object.type=MemberExpression][object.object.type=Identifier][object.object.name=process][object.property.type=Identifier][object.property.name=env][property.type=Identifier][property.name!=NODE_ENV]",
+                        message:
+                            "Environment variables other than NODE_ENV are not allowed in next.config.js. Use SiteConfig or a custom Context instead",
+                    },
+                ],
+            },
+        },
+        {
+            files: ["*.loader.ts"],
```

Comment:
@fraxachun loader.ts is just a convention: the file can have any name or the loader function can be defined in the component file

do you think this catches a common error, although not 100%?